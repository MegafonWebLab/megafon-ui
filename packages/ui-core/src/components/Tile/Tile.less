@import '~styles/base.less';

@b: e('.mfui-beta-tile');

@{b} {
    background-color: @clearWhite;

    &_radius_rounded {
        border-radius: 8px;
    }

    &_pointer {
        cursor: pointer;
    }

    &_theme_light&_shadow_low {
        box-shadow: 0 0.25px 3px rgba(0, 0, 0, 0.25);
    }

    &_theme_light&_shadow_high {
        box-shadow: 0 2px 9px rgba(0, 0, 0, 0.15);
    }

    &_theme_light&_shadow_hover {
        box-shadow: 0 8px 28px rgba(0, 0, 0, 0.1);
    }

    &_theme_dark&_shadow_low {
        box-shadow: 0 0.25px 3px rgba(0, 0, 0, 0.45);
    }

    &_theme_dark&_shadow_high {
        box-shadow: 0 2px 9px rgba(0, 0, 0, 0.25);
    }

    &_theme_dark&_shadow_hover {
        box-shadow: 0 8px 28px rgba(0, 0, 0, 0.2);
        cursor: pointer;
    }

    &__link {
        color: inherit;
    }

    &__link,
    &__link:hover {
        text-decoration: none;
    }
}
