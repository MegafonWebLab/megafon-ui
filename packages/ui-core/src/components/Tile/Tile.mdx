---
name: Tile
menu: Компоненты
route: components/Tile
---

import { Playground, Props } from 'docz';
import Tile from './Tile';
import { DemoTileWrapper, tileInner } from './Tile.docz';

# Tile

## Properties

<Props of={Tile} />

## Rounded and squared tiles on light background

<Playground>
    <DemoTileWrapper>
        <Tile shadowLevel="zero" theme="light">
            <div style={tileInner}>Zero shadow</div>
        </Tile>
        <Tile shadowLevel="low" theme="light">
            <div style={tileInner}>Low shadow</div>
        </Tile>
        <Tile shadowLevel="high" theme="light">
            <div style={tileInner}>High shadow</div>
        </Tile>
        <Tile shadowLevel="hover" theme="light">
            <div style={tileInner}>Hover shadow</div>
        </Tile>
    </DemoTileWrapper>
    <DemoTileWrapper>
        <Tile radius="rounded" shadowLevel="zero" theme="light">
            <div style={tileInner}>Zero shadow</div>
        </Tile>
        <Tile radius="rounded" shadowLevel="low" theme="light">
            <div style={tileInner}>Low shadow</div>
        </Tile>
        <Tile radius="rounded" shadowLevel="high" theme="light">
            <div style={tileInner}>High shadow</div>
        </Tile>
        <Tile radius="rounded" shadowLevel="hover" theme="light">
            <div style={tileInner}>Hover shadow</div>
        </Tile>
    </DemoTileWrapper>
</Playground>

## Rounded and squared tiles on dark background

<Playground>
    <DemoTileWrapper wrapperTheme="dark">
        <Tile shadowLevel="zero" theme="dark">
            <div style={tileInner}>Zero shadow</div>
        </Tile>
        <Tile shadowLevel="low" theme="dark">
            <div style={tileInner}>Low shadow</div>
        </Tile>
        <Tile shadowLevel="high" theme="dark">
            <div style={tileInner}>High shadow</div>
        </Tile>
        <Tile shadowLevel="hover" theme="dark">
            <div style={tileInner}>Hover shadow</div>
        </Tile>
    </DemoTileWrapper>
    <DemoTileWrapper wrapperTheme="dark">
        <Tile radius="rounded" shadowLevel="zero">
            <div theme="dark" style={tileInner}>Zero shadow</div>
        </Tile>
        <Tile radius="rounded" shadowLevel="low">
            <div theme="dark"style={tileInner}>Low shadow</div>
        </Tile>
        <Tile radius="rounded" shadowLevel="high">
            <div theme="dark"style={tileInner}>High shadow</div>
        </Tile>
        <Tile radius="rounded" shadowLevel="hover">
            <div theme="dark"style={tileInner}>Hover shadow</div>
        </Tile>
    </DemoTileWrapper>
</Playground>

## Tiles with links

<Playground>
    <DemoTileWrapper>
        <Tile shadowLevel="zero" href="#">
            <div style={tileInner}>Zero shadow</div>
        </Tile>
        <Tile shadowLevel="low" href="#">
            <div style={tileInner}>Low shadow</div>
        </Tile>
        <Tile shadowLevel="high" href="#">
            <div style={tileInner}>High shadow</div>
        </Tile>
        <Tile shadowLevel="hover" href="#">
            <div style={tileInner}>Hover shadow</div>
        </Tile>
    </DemoTileWrapper>
</Playground>

## Interactive tiles

<Playground>
    <DemoTileWrapper>
        <Tile isInteractive shadowLevel="zero">
            <div style={tileInner}>Zero shadow</div>
        </Tile>
        <Tile isInteractive shadowLevel="low">
            <div style={tileInner}>Low shadow</div>
        </Tile>
        <Tile isInteractive shadowLevel="high">
            <div style={tileInner}>High shadow</div>
        </Tile>
        <Tile isInteractive shadowLevel="hover">
            <div style={tileInner}>Hover shadow</div>
        </Tile>
    </DemoTileWrapper>
</Playground>
