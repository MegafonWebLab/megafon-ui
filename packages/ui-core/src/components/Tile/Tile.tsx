import React from 'react';
import PropTypes from 'prop-types';
import cnCreate from 'utils/cnCreate';
import './Tile.less';
import filterDataAttrs, { IDataAttributes } from './../../utils/dataAttrs';

export const Theme = {
    LIGHT: 'light',
    DARK: 'dark',
} as const;

type ThemeType = typeof Theme[keyof typeof Theme];

export const Radius = {
    DEFAULT: 'default',
    ROUNDED: 'rounded',
} as const;

type RadiusType = typeof Radius[keyof typeof Radius];

export const Shadow = {
    ZERO: 'zero',
    LOW: 'low',
    HIGH: 'high',
    HOVER: 'hover',
} as const;

type ShadowType = typeof Shadow[keyof typeof Shadow];

export interface ITileProps extends IDataAttributes {
    /** Ссылка */
    href?: string;
    /** Тема */
    theme?: ThemeType;
    /** Радиус границы */
    radius?: RadiusType;
    /** Уровень тени */
    shadowLevel?: ShadowType;
    /** Включить тень при наведении */
    isInteractive?: boolean;
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Обработчик клика */
    onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

const cn = cnCreate('mfui-beta-tile');
const Tile: React.FC<ITileProps> = ({
    href,
    children,
    className,
    theme = 'light',
    shadowLevel = 'zero',
    radius = 'default',
    isInteractive = false,
    onClick,
    dataAttrs,
}) => {
    const [currentShadow, setCurrentShadow] = React.useState(shadowLevel);

    const handleClick = (e: React.MouseEvent<HTMLDivElement>): void => {
        onClick && onClick(e);
    };

    const handleHoverTile = (): void => {
        isInteractive && setCurrentShadow(Shadow.HOVER);
    };

    const handleMouseLeave = (): void => {
        setCurrentShadow(shadowLevel);
    };

    return (
        <div
            className={cn(
                {
                    theme,
                    radius,
                    shadow: currentShadow,
                    pointer: !!href,
                },
                className
            )}
            onClick={handleClick}
            onMouseEnter={handleHoverTile}
            onMouseLeave={handleMouseLeave}
            {...filterDataAttrs(dataAttrs)}
        >
            {href && (
                <a href={href} className={cn('link')}>
                    {children}
                </a>
            )}
            {!href && children}
        </div>
    );
};

Tile.propTypes = {
    href: PropTypes.string,
    theme: PropTypes.oneOf(Object.values(Theme)),
    radius: PropTypes.oneOf(Object.values(Radius)),
    shadowLevel: PropTypes.oneOf(Object.values(Shadow)),
    className: PropTypes.string,
    isInteractive: PropTypes.bool,
    onClick: PropTypes.func,
    dataAttrs: PropTypes.objectOf(PropTypes.string.isRequired),
};

export default Tile;
