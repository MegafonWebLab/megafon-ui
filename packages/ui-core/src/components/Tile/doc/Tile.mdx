---
name: Tile
menu: Компоненты
route: components/Tile
---

import { Props } from 'docz';
import Tile from '../Tile';
import Example from './Tile.example.mdx';

# Tile

<CopyToClipboardBox text="import { Tile } from '@megafon/ui-core';" />

<DoczTabs>
    <Example />
    <Props of={Tile} />
</DoczTabs>
