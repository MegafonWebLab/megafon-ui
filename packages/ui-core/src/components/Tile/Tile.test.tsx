import * as React from 'react';
import { shallow } from 'enzyme';
import Tile, { ITileProps, Theme, Radius, Shadow } from './Tile';

const props: ITileProps = {
    className: 'test-class',
    href: 'test-link',
    theme: Theme.DARK,
    radius: Radius.ROUNDED,
    shadowLevel: Shadow.HIGH,
    onClick: jest.fn(),
};

const tileSelector = '.mfui-beta-tile';
const tileHoverSelector = 'mfui-beta-tile_shadow_hover';
describe('<Tile />', () => {
    describe('snapshots', () => {
        it('renders component', () => {
            const wrapper = shallow(<Tile>Some test content</Tile>);
            expect(wrapper).toMatchSnapshot();
        });

        it('renders component with props', () => {
            const wrapper = shallow(
                <Tile {...props}>Some test content</Tile>
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('it renders with data attributes', () => {
            const wrapper = shallow(<Tile dataAttrs={{ 'data-test': 'test', 'incorrect-attr': 'test' }} />);
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('handlers', () => {
        it('calls onClick', () => {
            const handleClick = jest.fn();

            const wrapper = shallow(<Tile onClick={handleClick}>Some test content</Tile>);

            wrapper.simulate('click');
            expect(handleClick).toBeCalled();
        });

        it('should render shadow on mouse enter', () => {
            const wrapper = shallow(<Tile isInteractive>Some test content</Tile>);
            wrapper.find(tileSelector).simulate('mouseenter');

            expect(wrapper).toMatchSnapshot();
        });

        it('should render shadow on mouse leave', () => {
            const wrapper = shallow(<Tile isInteractive>Some test content</Tile>);
            wrapper.find(tileSelector).simulate('mouseleave');

            expect(wrapper).toMatchSnapshot();
        });
    });
});
