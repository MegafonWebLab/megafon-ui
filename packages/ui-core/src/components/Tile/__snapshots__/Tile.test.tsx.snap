// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> handlers should render shadow on mouse enter 1`] = `
<div
  className="mfui-beta-tile mfui-beta-tile_theme_light mfui-beta-tile_radius_default mfui-beta-tile_shadow_hover"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Some test content
</div>
`;

exports[`<Tile /> handlers should render shadow on mouse leave 1`] = `
<div
  className="mfui-beta-tile mfui-beta-tile_theme_light mfui-beta-tile_radius_default mfui-beta-tile_shadow_zero"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Some test content
</div>
`;

exports[`<Tile /> snapshots it renders with data attributes 1`] = `
<div
  className="mfui-beta-tile mfui-beta-tile_theme_light mfui-beta-tile_radius_default mfui-beta-tile_shadow_zero"
  data-test="test"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;

exports[`<Tile /> snapshots renders component 1`] = `
<div
  className="mfui-beta-tile mfui-beta-tile_theme_light mfui-beta-tile_radius_default mfui-beta-tile_shadow_zero"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Some test content
</div>
`;

exports[`<Tile /> snapshots renders component with props 1`] = `
<div
  className="mfui-beta-tile mfui-beta-tile_theme_dark mfui-beta-tile_radius_rounded mfui-beta-tile_shadow_high mfui-beta-tile_pointer test-class"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <a
    className="mfui-beta-tile__link"
    href="test-link"
  >
    Some test content
  </a>
</div>
`;
